exchanges/stream/websocket.go

```diff
@@ -283,27 +321,139 @@ func (w *Websocket) Connect() error {
+		}
+		if len(subs) != 0 {
+			if err := w.SubscribeToChannels(nil, subs); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	if len(w.connectionManager) == 0 {
 		w.setState(disconnectedState)
-		return fmt.Errorf("%v Error connecting %w", w.exchangeName, err)
+		return fmt.Errorf("cannot connect: %w", errNoPendingConnections)
 	}
-	w.setState(connectedState)
 
-	if !w.IsConnectionMonitorRunning() {
-		err = w.connectionMonitor()
+	// multiConnectFatalError is a fatal error that will cause all connections to
+	// be shutdown and the websocket to be disconnected.
+	var multiConnectFatalError error
+
+	// TODO: Implement concurrency below.
+	for i := range w.connectionManager {
+		if w.connectionManager[i].Setup.GenerateSubscriptions == nil {
+			multiConnectFatalError = fmt.Errorf("cannot connect to [conn:%d] [URL:%s]: %w ", i+1, w.connectionManager[i].Setup.URL, errWebsocketSubscriptionsGeneratorUnset)
+			break
+		}
+
+		subs, err := w.connectionManager[i].Setup.GenerateSubscriptions() // regenerate state on new connection
 		if err != nil {
-			log.Errorf(log.WebsocketMgr, "%s cannot start websocket connection monitor %v", w.GetName(), err)
+			if errors.Is(err, asset.ErrNotEnabled) {
+				if w.verbose {
+					log.Warnf(log.WebsocketMgr, "%s websocket: %v", w.exchangeName, err)
+				}
+				continue // Non-fatal error, we can continue to the next connection
+			}
+			multiConnectFatalError = fmt.Errorf("%s websocket: %w", w.exchangeName, common.AppendError(ErrSubscriptionFailure, err))
+			break
+		}
+
+		if len(subs) == 0 {
+			// If no subscriptions are generated, we skip the connection
+			if w.verbose {
+				log.Warnf(log.WebsocketMgr, "%s websocket: no subscriptions generated", w.exchangeName)
+			}
+			continue
+		}
+
+		if w.connectionManager[i].Setup.Connector == nil {
+			multiConnectFatalError = fmt.Errorf("cannot connect to [conn:%d] [URL:%s]: %w ", i+1, w.connectionManager[i].Setup.URL, errNoConnectFunc)
+			break
+		}
+		if w.connectionManager[i].Setup.Handler == nil {
+			multiConnectFatalError = fmt.Errorf("cannot connect to [conn:%d] [URL:%s]: %w ", i+1, w.connectionManager[i].Setup.URL, errWebsocketDataHandlerUnset)
+			break
+		}
+		if w.connectionManager[i].Setup.Subscriber == nil {
+			multiConnectFatalError = fmt.Errorf("cannot connect to [conn:%d] [URL:%s]: %w ", i+1, w.connectionManager[i].Setup.URL, errWebsocketSubscriberUnset)
+			break
+		}
+
+		// TODO: Add window for max subscriptions per connection, to spawn new connections if needed.
+
+		conn := w.getConnectionFromSetup(w.connectionManager[i].Setup)
+
+		err = w.connectionManager[i].Setup.Connector(context.TODO(), conn)
+		if err != nil {
+			multiConnectFatalError = fmt.Errorf("%v Error connecting %w", w.exchangeName, err)
+			break
 		}
+
+		w.Wg.Add(1)
+		go w.Reader(context.TODO(), conn, w.connectionManager[i].Setup.Handler)
+
+		w.connections[conn] = &w.connectionManager[i]
```

Comment:
There is a degree of disconnect between `w.connections` and `w.connectionManager.` Is there a reason you don't try to assign them and remove them at the same time?
I see on Shutdown, you clear `w.connections` while leaving `w.connectionManager`, but I'm not sure I'm getting the distinction between the two. I took them both as more of a design of a sliceMap, so you can instant lookup connections via a map, but can quickly iterate over them in a slice

Reply:
I am not really understanding assigning and removing at the same time sorry. 

As for the shutdown I clear the connections because they will drop in shutdown and not need to be used, the connection manager holds a singular ConnectionCandidate (will be named to ConnectionWrapper now) you can think of it basically  as a URL holder with handlers (which I can name that as well) then from that you can generate and spawn any amount of connections based off that. The connections map is purely for subscription management specifically, but will only come into play when there are many connections holding different subscriptions that need to be scaled and managed. But If for now this is too much for this PR, I can drop that map entirely iterate over the manager itself and when I do upgrade the system to spawn connections based of subscriptions I can reinstate?

Reply:
Thank you for your details!
I was approaching the Connection slice and map as like a sliceMap configuration, a way to quickly look up a connection via map, while being able to iterate over all connections in a slice.
What you're saying is that isn't the case and in that sense, the assigning things at the same time, doesn't really apply.
On next review I'll try to understand your design with these details in mind and comment from there